cpp/arcticdb/column_store/column.hpp

```diff
@@ -708,8 +733,49 @@ class Column {
+
+        if (!left_input_column.is_sparse() && !right_input_column.is_sparse()) {
+            // Both dense, use std::transform over the shorter column to avoid going out-of-bounds
+            if (left_input_column.row_count() <= right_input_column.row_count()) {
+                std::transform(left_input_data.cbegin<left_input_tdt, false>(), left_input_data.cend<left_input_tdt, false>(), right_input_data.cbegin<right_input_tdt, false>(), output_it, [f = std::forward<functor>(f)](auto left_value, auto right_value) {
+                    return f(left_value, right_value);
+                });
+            } else {
+                std::transform(right_input_data.cbegin<left_input_tdt, false>(), right_input_data.cend<left_input_tdt, false>(), left_input_data.cbegin<right_input_tdt, false>(), output_it, [f = std::forward<functor>(f)](auto right_value, auto left_value) {
+                    return f(left_value, right_value);
+                });
+            }
+        } else if (left_input_column.is_sparse() && right_input_column.is_sparse()) {
+            // Both sparse, only project rows in the intersection of on-bits from both sparse maps
+            auto left_accessor = random_accessor<left_input_tdt>(&left_input_data);
+            auto right_accessor = random_accessor<right_input_tdt>(&right_input_data);
+            // TODO: experiment with more efficient bitset traversal methods
+            // https://github.com/tlk00/BitMagic/tree/master/samples/bvsample25
+            for (auto set_bit = output_column.sparse_map().first(); set_bit < output_column.sparse_map().end(); ++set_bit) {
+                *output_it++ = f(left_accessor.at(*set_bit), right_accessor.at(*set_bit));
+            }
+        } else if (left_input_column.is_sparse() && !right_input_column.is_sparse()) {
+            // One sparse, one dense. Use the enumerating forward iterator over the sparse column as it is more efficient than random access
+            auto right_accessor = random_accessor<right_input_tdt>(&right_input_data);
+            const auto right_column_row_count = right_input_column.row_count();
+            for (auto left_it = left_input_data.cbegin<left_input_tdt, true, true>();
+                 left_it != left_input_data.cend<left_input_tdt, true, true>() && left_it->idx() < right_column_row_count;
+                 ++left_it) {
+                *output_it++ = f(left_it->value(), right_accessor.at(left_it->idx()));
+            }
+        } else if (!left_input_column.is_sparse() && right_input_column.is_sparse()) {
+            // One sparse, one dense. Use the enumerating forward iterator over the sparse column as it is more efficient than random access
+            auto left_accessor = random_accessor<left_input_tdt>(&left_input_data);
+            const auto left_column_row_count = left_input_column.row_count();
+            for (auto right_it = right_input_data.cbegin<right_input_tdt, true, true>();
```

Comment:
Because of `enumeration`, the combinatorics mean that you would then have:

- `cbegin_sparse_enumerated`
- `cbegin_dense_enumerated`
- `cbegin_sparse_non_enumerated`
- `cbegin_dense_non_enumerated`

and similar for `begin`, `end`, and `cend`. I'll swap the bools for enums though so make it more readable at the calling sites.